### PR TITLE
Render flexible plots

### PR DESCRIPTION
**Pull request by mattseddon**
# 1/2 `main` <- this <- #2452

Related to #1757

This PR is the first major step in accommodating the data required to render flexible plots. 

I have updated our approach to check each `datapoint` in the plots data for an object named `dvc_data_version_info`. This object can currently contain a `field` or `filename`. Multi-source (flexible) plots are identified by > 1 variation in this data. 

After collecting all of the variations in the data we then work out if we need to and how to apply a different `strokeDash`/`shape` for each variation.

These generate an encoding update which is passed to the webview as part of the template.

image

Screen Shot 2022-09-15 at 3 13 04 pm

**Note:** The next step is to add the new items shown in the zoomed-in plot's legend into the plots tree in the sidebar.

In order to test manually you would need a repo with flexible plots and this branch: https://github.com/iterative/dvc/pull/8282

### CODE REVIEW

File: extension/src/cli/dvc/constants.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-export const MIN_CLI_VERSION = '2.24.0'
+export const MIN_CLI_VERSION = '2.0.0'
```
Comment (mattseddon) [inline extension/src/cli/dvc/constants.ts:1]: [I] Remove

File: extension/src/plots/model/collect.ts [hunk 2]
```diff
@@ -355,7 +355,20 @@ const collectDatapoints = (
   values: Record<string, unknown>[] = []
 ) => {
   for (const value of values) {
-    ;(acc[rev][path] as unknown[]).push({ ...value, rev })
+    const filename = (
```
Comment (mattseddon) [inline extension/src/plots/model/collect.ts:358]: [I] Cleanup this garbage

File: webview/src/plots/components/ZoomedInPlot.tsx [hunk 3]
```diff
@@ -28,7 +28,12 @@ export const ZoomedInPlot: React.FC<ZoomedInPlotProps> = ({
         {...merge(
           { ...cloneDeep(props) },
           {
-            spec: { encoding: { color: { legend: { disable: false } } } }
+            spec: {
+              encoding: {
```
Comment (mattseddon) [inline webview/src/plots/components/ZoomedInPlot.tsx:32]: [I] This should be a function imported from the extension

File: extension/src/plots/model/collect.ts [hunk 4]
```diff
@@ -413,6 +426,46 @@ export const collectData = (
...
+    )?.filename
+    if (filename) {
+      files.add(filename)
+    }
+    if (files.size > 1) {
+      acc[path] = true
+      return
+    }
+  }
+}
```
Comment (pared) [inline extension/src/plots/model/collect.ts:451]: Both here and in `getSpecEncodingUpdate` we need to remember that that filename is not the only field differentiating "normal" plot from "flexible". We need to count "field" occurrences too - one might make a plot out of two columns from single `csv` file for example. Take a look at `isMultiSourcePlot` from 
https://github.com/iterative/vscode-dvc/pull/2379.

**Review by pared — CHANGES_REQUESTED** [review]

File: extension/src/plots/model/collect.ts (hunk 4 above)
Comment (mattseddon) [reply extension/src/plots/model/collect.ts:451]: Thank you, this was just garbage to be able to show something.

Comment (pared) [reply extension/src/plots/model/collect.ts:451]: Sure thing. I guess that's where marking particular points in different style might come in handy

File: extension/src/plots/model/collect.ts [hunk 5]
```diff
@@ -355,7 +355,20 @@ const collectDatapoints = (
...
+    )?.filename
+    const data: { rev: string; filename?: string } = {
+      ...value,
+      rev
+    }
+
+    if (filename) {
+      data.filename = filename
+    }
+    ;(acc[rev][path] as unknown[]).push(data)
```
Comment (wolmir) [inline extension/src/plots/model/collect.ts:371]: What about
```
const revisionPathDataItems = acc[rev][path] as unknown[]
revisionPathDataItems.push(data)
```
?

File: extension/src/cli/dvc/constants.ts (hunk 1 above)
Comment (mattseddon) [inline extension/src/cli/dvc/constants.ts:1]: [C] Cannot merge and ship before new version drops (https://github.com/iterative/dvc/pull/8282)

File: extension/src/plots/model/collect.ts [hunk 6]
```diff
@@ -487,18 +501,48 @@ export const collectTemplates = (data: PlotsOutput): TemplateAccumulator => {
...
+  const fields = splitConcatenatedFields(field)
+  return JSON.parse(
+    template.replace(
+      '"<DVC_METRIC_DATA>"',
+      JSON.stringify(
+        datapoints.map(data => {
+          const obj = data as Record<string, unknown>
+          return {
+            ...obj,
+            [field]: joinFields(fields.map(field => obj[field] as string))
```
Comment (mattseddon) [inline extension/src/plots/model/collect.ts:524]: [F] There is a situation where we concatenate fields together for the stoke dash update. If that happens we need to do the same thing with the data.

File: extension/src/plots/multiSource/collect.test.ts [hunk 7]
```diff
@@ -0,0 +1,168 @@
+import { join } from 'path'
+import { collectMultiSourceEncoding } from './collect'
+
+describe('collectMultiSourceEncoding', () => {
```
Comment (mattseddon) [inline extension/src/plots/multiSource/collect.test.ts:4]: [F] This should give simple examples of how the encoding update is collected for variations in the datapoints.

File: extension/src/plots/multiSource/constants.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
+export const StrokeDash = [
+  [1, 0],
+  [8, 8],
+  [8, 4],
+  [4, 4],
+  [4, 2],
+  [2, 1],
+  [1, 1]
```
Comment (mattseddon) [inline extension/src/plots/multiSource/constants.ts:8]: Lifted from dropdown here: https://vega.github.io/vega/docs/marks/line/

File: extension/src/plots/multiSource/constants.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+] as const
+export type StrokeDashValue = typeof StrokeDash[number]
+
+export const Shape = ['square', 'circle', 'triangle'] as const
+export type ShapeValue = typeof Shape[number]
```
Comment (mattseddon) [inline extension/src/plots/multiSource/constants.ts:13]: Same from here: https://vega.github.io/vega/docs/marks/symbol/

File: extension/src/plots/vega/util.test.ts [hunk 10]
```diff
@@ -264,3 +265,44 @@ describe('extendVegaSpec', () => {
     expect(updatedSpecString).toContain(truncatedTitle)
   })
 })
+
+describe('reverseOfLegendSuppressionUpdate', () => {
+  it('should reverse the legend suppression applied by extendVegaSpec', () => {
+    type NonOptionalEncoding = { [P in keyof Encoding]-?: Encoding[P] }
```
Comment (mattseddon) [inline extension/src/plots/vega/util.test.ts:271]: [F] This means that if the Encoding property is updated then compilation will fail unless this test is also updated.

Comment (mattseddon) [reply extension/src/plots/vega/util.test.ts:271]: Which should mean we don't miss the required webview update.

File: extension/src/test/fixtures/plotsDiff/index.ts [hunk 11]
```diff
@@ -480,8 +480,10 @@ const extendedSpecs = (plotsOutput: TemplatePlots): TemplatePlotSection[] => {
           } as TopLevelSpec,
           PlotSize.REGULAR,
           {
-            domain: expectedRevisions,
-            range: copyOriginalColors().slice(0, 5)
+            color: {
+              domain: expectedRevisions,
+              range: copyOriginalColors().slice(0, 5)
+            }
```
Comment (mattseddon) [inline extension/src/test/fixtures/plotsDiff/index.ts:486]: [F] We will need integration tests later.

File: extension/src/plots/model/collect.ts [hunk 12]
```diff
@@ -487,18 +501,42 @@ export const collectTemplates = (data: PlotsOutput): TemplateAccumulator => {
...
+  const fields = unmergeConcatenatedFields(field)
+  return JSON.parse(
+    template.replace(
+      '"<DVC_METRIC_DATA>"',
+      JSON.stringify(
+        datapoints.map(data => {
+          const obj = data as Record<string, unknown>
+          return {
+            ...obj,
+            [field]: mergeFields(fields.map(field => obj[field] as string))
```
Comment (mattseddon) [inline extension/src/plots/model/collect.ts:524]: [F] There is a situation where we concatenate fields together for the stoke dash update. If that happens we need to do the same thing with the data.

**Comment by julieg18** [comment]
Excited to test this out! 

> In order to test manually you would need a repo with flexible plots and this branch: https://github.com/iterative/dvc/pull/8282

I'm not very familiar with DVC plots. Does anyone know a repo with flexible plots? Or is there a way to alter a DVC repo to product flexible plots?

**Comment by pared** [comment]
@julieg18 
I usually create them via some script:
```bash
#!/bin/bash

set -exu

wsp=test_wspace
rep=test_repo

rm -rf $wsp && mkdir $wsp && pushd $wsp
main=$(pwd)

mkdir $rep && pushd $rep

git init
dvc init

mkdir subdir && pushd subdir

git add -A
git commit -am "initial"

echo -e "some_param: 0.1
" > params.yaml


echo -e "
import yaml
import json
import random

def get_params():
    with open('params.yaml', 'r') as fd:
        return yaml.safe_load(fd)
p = get_params()['some_param']


def prepare_cm(num_classes=3,num_samples=3):
    result = []
    for i in range(num_classes):
        for j in range(num_samples):
            if random.random() <= p:
                result.append({'actual': i, 'predicted': i})
            else:
                result.append({'actual': i, 'predicted': random.choice(list(range(num_samples)))})
    return result

res1=[]
res2=[]
for i in range(1, 11):
    res1.append({'x': i, 'y': i*p})
    res2.append({'x': i, 'z': i*p*2})

with open('res1.json', 'w') as fd:
    json.dump(res1, fd)

with open('res2.json', 'w') as fd:
    json.dump(res2, fd)

with open('metric1.json', 'w') as fd:
    json.dump(res1[-1],fd)

with open('metric2.json', 'w') as fd:
    json.dump(res2[-1],fd)

with open('confusion_matrix.json', 'w') as fd:
    json.dump(prepare_cm(), fd)

import shutil
shutil.copyfile('confusion_matrix.json', 'confusion_matrix2.json')
" > train.py

dvc run -d train.py -p some_param \
--plots-no-cache confusion_matrix.json \
--plots-no-cache res1.json \
-O res2.json -M metric1.json -M metric2.json \
-O confusion_matrix2.json -n train "python train.py"

echo -e "plots:
  res2.json:
    template: simple
  combined:
    x: x
    y:
      res1.json: y
      res2.json: y
  confusion_matrix2.json:
    template: confusion
    x: actual
    y: predicted" >> dvc.yaml
 

dvc plots modify res1.json --template scatter -x x -y y
dvc plots modify confusion_matrix.json --template confusion -x actual -y predicted

git add -A
git commit -am "initial run"

echo -e "some_param: 0.3
" > params.yaml

dvc exp run -n "p0.9"

echo -e "some_param: 0.9
" > params.yaml

dvc repro
```

Flexible plots are the ones that are created via top level plot definition in `dvc.yaml` (in given script its the code appending plots section to `dvc.yaml`).

**Comment by sroy3** [comment]
A little CSS bug:

https://user-images.githubusercontent.com/3683420/192293426-9c69e980-a737-4db9-b611-13312236a57f.mov

**Comment by mattseddon** [comment]
> A little CSS bug:
> 
>  Screen.Recording.2022-09-26.at.9.48.09.AM.mov 
> Same thing happening in a non-zoomed-in-plot:
> 
>  Screen.Recording.2022-09-26.at.9.49.44.AM.mov

Is this new or existing?

**Comment by sroy3** [comment]
> > A little CSS bug:
> > Screen.Recording.2022-09-26.at.9.48.09.AM.mov
> > Same thing happening in a non-zoomed-in-plot:
> > Screen.Recording.2022-09-26.at.9.49.44.AM.mov
> 
> Is this new or existing?

Not sure, can only reproduce with the flexible plot. That is because the tooltip goes above the plot. I couldn't find any other plot that has a point that high.

**Comment by julieg18** [comment]
> A little CSS bug:

> Is this new or existing?

Appears to be existing since I can replicate in an `example-get-started` plot: 

https://user-images.githubusercontent.com/43496356/192388065-0df6a2a4-7f0d-4bca-8b9c-ecbdbbb542f1.mov

**Comment by mattseddon** [comment]
Will follow up on the css bug separately.

**Review by sroy3 — APPROVED** [review]

**Pull request merged**